web-api/src/storage/elastic.rs

```diff
@@ -186,39 +192,94 @@ impl Client {
-                        .map_err(Into::into),
-                    serde_json::to_value(IngestedDocument {
+        let pre_deletions = documents
+            .iter()
+            .filter_map(|document| (document.embeddings.len() > 1).then_some(&document.id))
+            .collect_vec();
+
+        self.delete_by_parents(pre_deletions).await?;
+
+        let bulk = documents
+            .iter()
+            .flat_map(|document: &models::IngestedDocument| {
+                if document.embeddings.len() == 1 {
+                    let index = BulkInstruction::Index {
+                        id: document.id.as_str().to_owned(),
+                    };
+                    let data = EsDocument {
                         snippet: &document.snippet,
                         properties: &document.properties,
-                        embedding: &document.embedding,
+                        embedding: document.embeddings.first().unwrap(),
                         tags: &document.tags,
-                    })
-                    .map_err(Into::into),
-                ]
-            }))
-            .await?;
+                        parent: None,
+                    };
+                    vec![
+                        serde_json::to_value(index).map_err(Into::into),
+                        serde_json::to_value(data).map_err(Into::into),
+                    ]
+                } else {
+                    document
+                        .embeddings
+                        .iter()
+                        .enumerate()
+                        .flat_map(|(idx, embedding)| {
+                            let index = BulkInstruction::Index {
+                                id: document.id.create_child_id(idx),
+                            };
+                            let data = EsDocument {
+                                snippet: &document.snippet,
+                                properties: &document.properties,
+                                embedding,
+                                tags: &document.tags,
+                                parent: Some(&document.id),
+                            };
+                            vec![
+                                serde_json::to_value(index).map_err(Into::into),
+                                serde_json::to_value(data).map_err(Into::into),
+                            ]
+                        })
+                        .collect()
+                }
+            });
+
+        let response = self.bulk_request(bulk).await?;
         Ok(response.failed_documents("index", false).into())
     }
 
+    pub(super) async fn delete_by_parents(&self, parents: Vec<&DocumentId>) -> Result<(), Error> {
+        let url = self.create_url(["_delete_by_query"], [("refresh", None)]);
+        let body = json!({
+            "query": {
+                "terms": {
+                    "parent": parents,
+                }
```

Comment:
:thinking:  In all cases we do have the number of childs the new document has. We could use this to instead of running "delete where `parent={parent}`" run "delete where `parent={parent} && child_idx >= {child_count}`.

Not sure if it's worth the additional index field.